optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/locator/EqualsLocationStrategy.java

```diff
@@ -22,7 +22,11 @@
 
     @Override
     public void addWorkingObject(Map<Object, Object> idToWorkingObjectMap, Object workingObject) {
-        idToWorkingObjectMap.put(workingObject, workingObject);
+        Object oldAddedObject = idToWorkingObjectMap.put(workingObject, workingObject);
+        if (oldAddedObject != null) {
+            throw new IllegalStateException("An object that is equal to workingObject (" + workingObject
+                    + ") has already been added.");
+        }
```

Comment:
I think it is highly improbable there would be a difference in the two equal object that would affect toString result. But yes, an exception is an exceptional event and some extra information can't do any harm.